lints/type_cosplay/src/lib.rs

```diff
@@ -61,31 +64,63 @@ struct TypeCosplay {
     deser_types: FxHashMap<AdtKind, Vec<(DefId, Span)>>,
 }
 
+// get type X
+// check if implements Discriminator
+// check corresponding function call type:
+// if !try_deserialize
+// emit lint with warning to use try_deserialize (because any types deriving Discriminator should since it guaranteed to check discrim)
+
 impl<'tcx> LateLintPass<'tcx> for TypeCosplay {
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
         if_chain! {
             if !expr.span.from_expansion();
-            if let ExprKind::Call(fnc_expr, args_exprs) = expr.kind;
-            if is_deserialize_function(cx, fnc_expr);
+            if let ExprKind::Call(fnc_expr, _args_exprs) = expr.kind;
+            // TODO: is the following if statement really needed?? don't think it's ever used.
+            // all it does is check if AccountInfo.data is referenced...but what bytes we deser
+            // from shouldn't impact if this is a type-cosplay issue or not.
             // walk each argument expression and see if the data field is referenced
-            if args_exprs.iter().any(|arg| {
-                visit_expr_no_bodies(arg, |expr| contains_data_field_reference(cx, expr))
-            });
```

Comment:
I think this check is still needed to ensure we are deserializing from `AccountInfo.data`.